### DISCUSSION TITLE
How do you conditionally generate code?

### DISCUSSION THREAD

**Discussion by farhansolodev** [GoogleFeud/ts-macros#80] [Q&A] [answered]
I want to generate code for every property of a class Foo except for Foo's methods. I am using `$$propsOfType` within repetitions to do this, but since that macro also returns the classes methods, I want to skip every run of a repetition where the property returned by `$$propsOfType` does not exist in `Object.keys(Foo)` since `Object.keys` does not return a classes methods, which is want I want.

To be clear, this:
```ts
    +[[$$propsOfType!<F>()], (prop) => {
        $$ts!(`${toVarName}.${prop} = ${fromVarName}.${prop}`)
    }];
```
generates this:
```ts
        this.prop1 = prop.prop1;
        this.prop2 = prop.prop2;
        this.prop3 = prop.prop3;
        this.prop420 = prop.prop420;
        this.prop1337 = prop.prop1337;
        this.someMethod = prop.someMethod; // dont want this
```
but I'd like to do something like:
```ts
    +[[$$propsOfType!<F>()], (prop) => {
        if (Object.keys(F).includes(prop)) { // just pseudocode, does not compile
        	$$ts!(`${toVarName}.${prop} = ${fromVarName}.${prop}`)
        }
    }];
```
to generate:
```ts
        this.prop1 = prop.prop1;
        this.prop2 = prop.prop2;
        this.prop3 = prop.prop3;
        this.prop420 = prop.prop420;
        this.prop1337 = prop.prop1337;
        // this.someMethod = prop.someMethod; // this is now not generated
```

**Comment by volenslavchev** [comment, marked as answer]
Hi! You can do this with the `$$typeMetadata` built-in macro, which separates the object's properties and methods:

```ts
function $test<T>(toName: string, fromName: string) {
    +[[$$typeMetadata!<T>(true).properties], (prop: TypeMetadataProperty) => {
        $$ts!(`${toName}.${prop.name} = ${fromName}.${prop.name}`);
    }]
}
```

[playground](https://googlefeud.github.io/ts-macros/?code=FAMwrgdgxgLglgewgAgCQwKYGcYB4AqAfABQwIByAhgLYYBcyOATnBAOYA0yITC1VtBs1ZsAlMgDewZDOQBqANoLU6AJ4AHDAFkMMSgBNKegIQESMJmAyiAdOt6am8bAF0uxewnUN8G7boMjSgAFBwwnVXEAXkJJaVkElRgsY2IAA1QJMgEMAF8bTM91GwgaPOQotAkePhz8wocSsty00QBueJlcl2Bc4GAoABtKLCxkLVUAYWHRuITKIQsRDoSAIwYIMGpV8JXZKAZVhARBjEoIPZkoJGYwWAQmYkM9BgAlDGumfVxhdi5IADWEAQAHcIIRxFIEolMDhTBNpiMsCQAEQwAAWcCwKK4KOelBR7U6yD6xIxImIkOJCSYujATBQAEZLiTekA)

**Reply by farhansolodev** [reply]
You sir are a legend!
